EIPS/eip-ase.md

```diff
@@ -0,0 +1,88 @@
+---
+title: EOF - Prepare for Address Space Extension
+description: Update EOF opcodes so addresses are not trimmed durring execution
+author: Danno Ferrin (@shemnon)
+discussions-to: tbd
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-04-03
+requires: 3540, 3670, 7069
+---
+
+## Abstract
+
+Operations in the Legacy EVM trim off the top 12 bytes of an address operand before evaluation. This
+EIP changes the handling of those opcodes within EOF so that no trimming occurs and the top twelve
+bytes need to be zero or an exceptional halt is raised.
+
+## Motivation
+
+There have been propsals to extend Ethereum Addresses from 160 bytes to 256, such as one that
+would [use the extra bits for state expiry](https://ethereum-magicians.org/t/increasing-address-size-from-20-to-32-bytes/5485).
+One issue ground this work to a halt: EVM opcodes that accept Addresses trim all but the lowest 20
+bytes out fo the operand before processing. EVM Reference
+tests [verify this behavior](https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stBadOpcode/invalidAddrFiller.yml).
+
+The EVM Object Framework presents an opportunity to remove this address masking in a backwards
+compatible way, by baking it into the format definition from the start.
+
+Most of the work is already in place. The following 5 operations have already been banned from
+EOF: `EXTCODESIZE`, `EXTCODECOPY`, `EXTCODEHASH`, `CALLCODE`, and `SELFDESTRUCT`. Three call
+operations, `CALL`, `DELEGATECALL`, and `STATICCALL` are being revamped
+in [EIP-7069](./eip-7069.md). That leaves only one operation, `BALANCE`, to be changed.
+
+When future uses of address space extension are specified it is expected that the exeptional halt
+behavior will be modified.
+
+## Specification
+
+The `BALANCE` operation, when invoked in code in an EOF container, will reauire the top 12 bytes of
```

Comment:
An interesting side effect is that this provides a cheap (in terms of codesize, not gas) way of checking that there are no high bits set in a stack item.

Reply:
Cheap?  It's going to get hit with a cold-account charge on success - https://eips.ethereum.org/EIPS/eip-2929 - and a revert on failure.  I thing PUSH1 0x20 SHR ISZERO would be cheaper.

Reply:
"cheap" - as in smaller bytecode. i don't think it is that practical unless you have really strong codesize constraints, i'm just pointing out that it's a new side effect.